### DISCUSSION TITLE
How to downgrade to Preview 14?

### DISCUSSION THREAD

**Discussion by emorell96** [dotnet/maui#6059] [Q&A] [answered]
How can I downgrade to preview 14? RC1 is just broken, and I cannot even build with MSAL which was working since preview 10! How can I downgrade Visual Studio?

**Comment by jsuarezruiz** [comment]
First of all, please, fill an issue with more details!. We want to track and review any possible issue like this. 

You can downgrande, uninstalling and intalling the correct workloads. For that, you can use  https://github.com/Redth/dotnet-maui-check/blob/main/Clean-Old-DotNet6-Previews.ps1 to uninstall the current installed version.

Then, use this script:

```
iwr https://aka.ms/dotnet/maui/maui-install.ps1 -OutFile maui-install.ps1;
.\maui-install.ps1 -b 'release/6.0.2xx-preview14' -v '6.0.200-preview'
```

To install Preview 14. You can find more details here: https://github.com/dotnet/maui/blob/main/.github/DEVELOPMENT.md

**Reply by emorell96** [reply]
I have opened an issue before: https://github.com/dotnet/maui/issues/6052

Thank you for the instructions because this RC1 has taken me back to 0 by breaking my app build.

**Reply by emorell96** [reply]
It just feels like this RC1 has gone in the wrong direction. I just see how many new issues are being reported seems way more than the issues that were opened on Preview 14's release.

**Reply by emorell96** [reply]
@jsuarezruiz would you know why this fails?
I am trying to go back to preview 14 but I get this error:
```
Installation rollback failed: One or more errors occurred. (microsoft.netcore.app.runtime.mono.android-arm.msi.x64::6.0.3 is not found in NuGet feeds https://pkgs.dev.azure.com/dnceng/public/_packaging/darc-pub-dotnet-runtime-3ba4b3f4/nuget/v3/index.json;https://pkgs.dev.azure.com/dnceng/public/_packaging/darc-pub-dotnet-emsdk-ee0a97a0-1/nuget/v3/index.json;https://pkgs.dev.azure.com/dnceng/public/_packaging/darc-pub-dotnet-emsdk-ee0a97a0/nuget/v3/index.json;https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet-eng/nuget/v3/index.json;https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet-public/nuget/v3/index.json;https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet6/nuget/v3/index.json;https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet7/nuget/v3/index.json;https://pkgs.dev.azure.com/xamarin/public/_packaging/SkiaSharp/nuget/v3/index.json".)
```

**Reply by emorell96** [reply]
@jsuarezruiz do you know how to get the .NET SDK 6.0.300 ? I am trying to test a few of the bugs I found in RC1, so I am following this: https://github.com/dotnet/maui/blob/main/.github/DEVELOPMENT.md#rc-1-branch

But I only found 6.0.202 in https://dotnet.microsoft.com/en-us/download/dotnet/6.0

Thanks

**Comment by emorell96** [comment, marked as answer]
I was able to downgrade but there's a few gotchas that I'd like to share if anyone comes across this post.

1. Use this script https://github.com/Redth/dotnet-maui-check/blob/main/Clean-Old-DotNet6-Previews.ps1 shared by Javier.
2. First Gotcha: You will need Powershell 7 to use it: https://docs.microsoft.com/en-us/powershell/scripting/install/installing-powershell-on-windows?view=powershell-7.2#msi
3. Once you've done that I tried to run:
```
iwr https://aka.ms/dotnet/maui/maui-install.ps1 -OutFile maui-install.ps1;
.\maui-install.ps1 -b 'release/6.0.2xx-preview14' -v '6.0.200-preview'
```
But it complained that there wasn't any .NET SDK. So I had to install the right sdk before this step. For Preview 14 you will need the **.NET 6.02 SDK**: https://dotnet.microsoft.com/en-us/download/dotnet/6.0
4. Clear your nuget cache:
```
dotnet nuget locals all --clear
```

5. Once this is installed I could run the above command but with a modification if you want to use the MAUI community toolkit ver pre8:
```
iwr https://aka.ms/dotnet/maui/maui-install.ps1 -OutFile maui-install.ps1;
.\maui-install.ps1 -b '6.0.200-preview.14.2' -v '6.0.200-preview'
```
The last line is important because MAUI Community Toolkit requires Maui preview 14.5099 but if you run the original command it will install Maui Preview 14.5097.

Hope this helps anyone who might have the same issues I had with RC1 which makes building an app almost impossible.

---------

If Powershell complains about the script not being digitally signed use:
```
Set-ExecutionPolicy -Scope Process -ExecutionPolicy Bypass
```

This allows you to bypass the protection only for that PowerShell window.

**Reply by pictos** [reply]
@emorell96 we just released a new preview version of the CommunityToolkit.Maui that works with MAUI RC1, if you want to try

**Reply by emorell96** [reply]
Thank you @pictos, I tried it, sadly the MAUI community toolkit worked perfectly, the issue I have is the MAUI RC1, specially this bug: https://github.com/dotnet/maui/issues/6067 that makes the app I am developing right now unusable. So until this gets fixed or at least a workaround is found I am staying in Preview 14.

**Reply by icarrenomercado** [reply]
You can try this handler to load remote images until they release RC3:

    ImageHandler.Mapper.AppendToMapping(nameof(ImageView.Drawable), async (handler, view) =>
    {
        if (view.Source is UriImageSource uriImageSource)
            try
            {
                byte[] imageData;
                using (var response = await _httpClient.GetAsync(uriImageSource.Uri))
                {
                    imageData = await response.Content.ReadAsByteArrayAsync();
                }
                handler.PlatformView.SetImageDrawable(new BitmapDrawable(BitmapFactory.DecodeByteArray(imageData, 0, imageData.Length)));
            }
            catch
            {
            }
        ;
    });

**Reply by dharber** [reply]
Granted RC3 is out now but where would this code have resided? I'm assuming in some Android platform code but I don't really understand handlers well enough to figure out where it would be invoked.

**Reply by JeffIPV** [reply]
For anyone on RC3 that wants to revert back to RC2.
run cmd pmrpt as administrator and run `dotnet workload install maui --from-rollback-file https://aka.ms/dotnet/maui/rc.2.json --source https://aka.ms/dotnet6/nuget/index.json --source https://api.nuget.org/v3/index.json --include-previews`
or for mac sudo dotnet ......

**Comment by dharber** [comment]
Sadly, not only did this fail to install for me, it broke the existing installation so I now have to reinstall VS. (There's probably a way to recover it but I have no idea what it is.)

So, while it's an interesting prospect, theoretically, tread carefully.
